python/ray/autoscaler/v2/instance_manager/reconciler.py

```diff
@@ -106,6 +107,52 @@ def reconcile(
         )
         return autoscaling_state
 
+    @staticmethod
+    def initialize_head_node(
+        instance_manager: InstanceManager,
+        non_terminated_cloud_instances: Dict[CloudInstanceId, CloudInstance],
+    ):
+        """
+        Initialize the head node for the instance manager.
+
+        This is needed since the head node is not managed by the instance manager
+        as of now and needs to be initialized separately.
+
+        This should be called when the autoscaler is first started.
+
+        Args:
+            instance_manager: The instance manager to reconcile.
+            non_terminated_cloud_instances: The non-terminated cloud instances from
+                the cloud provider.
+        """
+        head_cloud_instance = None
+        for cloud_instance in non_terminated_cloud_instances.values():
+            if cloud_instance.node_kind == NodeKind.HEAD:
+                head_cloud_instance = cloud_instance
+                break
+
+        assert head_cloud_instance, (
+            "Head node is not found. Make sure the head node "
+            "is started with RAY_CLOUD_INSTANCE_ID set, and it has the needed labels "
+            "specified in ray/autoscaler/tags.py ."
+        )
+
+        # Initialize the head node for the instance manager.
+        updates = {}
+        head_node_instance_id = InstanceUtil.random_instance_id()
+        updates[head_node_instance_id] = IMInstanceUpdateEvent(
```

Comment:
So this will create and update the status to ALLOCATED in one go?

Reply:
Yes - see this PR for the instance manager change for upsert: https://github.com/ray-project/ray/pull/43098

Reply:
Explicitly set upsert=true for cases where upsert is expected.